src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,126 @@
 package core.basesyntax;
 
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int INITIAL_CAPACITY = 16;
+    private static final int GROW_FACTOR = 2;
+    private static final float LOAD_FACTOR = 0.75f;
+    private Object[] entries;
+    private int size;
+    private int threshold;
+
+    public MyHashMap() {
+        entries = new Object[INITIAL_CAPACITY];
+        threshold = (int)(INITIAL_CAPACITY * LOAD_FACTOR);
+    }
 
     @Override
     public void put(K key, V value) {
-
+        int index = getIndex(key);
+        if (entries[index] == null) {
+            addNewEntry(key, value, index);
+        } else {
+            addOrUpdateEntry(key, value, index);
+        }
+        if (size > threshold) {
+            resize();
+        }
     }
 
     @Override
     public V getValue(K key) {
+        int index = getIndex(key);
+        if (entries[index] == null) {
+            return null;
+        }
+        return getValue(key, index);
+    }
+
+    @SuppressWarnings("unchecked")
+    private V getValue(K key, int index) {
+        Entry<K, V> entry = (Entry<K, V>) entries[index];
+        while (true) {
+            if (checkKeyEquality(key, entry.key)) {
+                return entry.value;
+            }
+            if (entry.next == null) {
+                break;
+            } else {
+                entry = entry.next;
+            }
+        }
         return null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private int getIndex(K key) {
+        return Math.abs(getHashCode(key) % entries.length);
+    }
+
+    private int getHashCode(K key) {
+        return key == null ? 0 : key.hashCode();
+    }
+
+    private void addNewEntry(K key, V value, int index) {
+        entries[index] = new Entry<>(key, value);
+        size++;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void addOrUpdateEntry(K key, V value, int index) {
+        Entry<K, V> entryAtIndex = (Entry<K, V>) entries[index];
+        while (true) {
+            if (checkKeyEquality(key, entryAtIndex.key)) {
+                entryAtIndex.value = value;
+                return;
+            }
+            if (entryAtIndex.next == null) {
+                break;
+            } else {
+                entryAtIndex = entryAtIndex.next;
+            }
```

Comment:
Same

Reply:
I can't simplify this block of code, because i need to check before assignment  `entryAtIndex = entryAtIndex.next` if it will be null, if it will i break, and my variable entryAtIndex points to the last element in bucket. If i will make this line as you suggested, at the end my variable entryAtIndex will be null, and after the while loop i ca'nt do like i'm doing: entryAtIndex.next = new Entry<>(key, value).